app/src/main/java/com/celzero/bravedns/service/BraveVPNService.kt

```diff
@@ -269,86 +263,157 @@ class BraveVPNService : VpnService(), ConnectionCapabilityMonitor.NetworkListene
+    private fun blockBackgroundData(): Boolean {
+        if (!persistentState.isBackgroundEnabled) return false
+        if (!accessibilityServiceRunning()) return false
+        return true
+    }
+
+    private fun accessibilityServiceRunning(): Boolean {
+        val now = elapsedRealtime()
+        if (Math.abs(now - accessibilityHearbeatTimestamp) > Constants.ACCESSIBILITY_SERVICE_HEARTBEAT_THRESHOLD_MS) {
+            accessibilityHearbeatTimestamp = now
+            isAccessibilityServiceRunning = Utilities.isAccessibilityServiceEnabledEnhanced(this, BackgroundAccessibilityService::class.java)
+        }
+        return isAccessibilityServiceRunning
+    }
+
+    private fun notifyLoadFailure() {
+        if (isLoadFailureNotificationActive()) {
+            val mainActivityIntent = PendingIntent.getActivity(this, 0, Intent(this, HomeScreenActivity::class.java), PendingIntent.FLAG_UPDATE_CURRENT)
+            var builder: NotificationCompat.Builder
+            if (VERSION.SDK_INT >= VERSION_CODES.O) {
+                val name: CharSequence = NOTIF_LOAD_RULES_FAIL
+                val description = resources.getString(R.string.rules_load_failure_heading)
+                val importance = NotificationManager.IMPORTANCE_HIGH
+                val channel = NotificationChannel(NOTIF_LOAD_RULES_FAIL, name, importance)
+                channel.description = description
+                notificationManager.createNotificationChannel(channel)
+                builder = NotificationCompat.Builder(this, NOTIF_LOAD_RULES_FAIL)
+            } else {
+                builder = NotificationCompat.Builder(this, NOTIF_LOAD_RULES_FAIL)
+            }
+
+            val contentTitle = resources.getString(R.string.rules_load_failure_heading)
+            val contentText = resources.getString(R.string.rules_load_failure_desc)
+            builder.setSmallIcon(R.drawable.dns_icon)
+                .setContentTitle(contentTitle)
+                .setContentIntent(mainActivityIntent)
+                .setContentText(contentText)
+            builder.setStyle(NotificationCompat.BigTextStyle().bigText(contentText))
+            builder.color = ContextCompat.getColor(this, getThemeAccent(this))
+            val openIntent = makeVpnIntent(this, NOTIF_ID_LOAD_RULES_FAIL, Constants.NOTIF_ACTION_RULES_FAILURE)
+            val notificationAction: NotificationCompat.Action = NotificationCompat.Action(0, resources.getString(R.string.rules_load_failure_reload), openIntent)
+            builder.addAction(notificationAction)
+
+            // Secret notifications are not shown on the lock screen.  No need for this app to show there.
+            // Only available in API >= 21
+            builder = builder.setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
+
+            builder.build()
+            notificationManager.notify(NOTIF_ID_LOAD_RULES_FAIL, builder.build())
+        }
+    }
+
+    /**
+     * Checks if the notification is already available in the status bar.
```

Comment:
It will return `false` once the user dismisses the notification. Also the firewall rules are getting loaded even before the service goes to started state every-time. Maybe we can wait until we hit this notification once?

Reply:
This has been addressed to avoid the scenario noted above.